File: day5/day5.py
from functools import cmp_to_key


def parse_input(filename):
    with open(filename, "r") as f:
        lines = f.read()
        tmp = lines.split("\n\n")
        rules = tmp[0].split("\n")
        updates_str = tmp[1].split("\n")

        # finding smaller and bigger number (1-step)
        smaller_numbers = {}
        bigger_numbers = {}
        for rule in rules:
            tmp = rule.split("|")
            left_num = int(tmp[0])
            right_num = int(tmp[1])

            if left_num not in bigger_numbers:
                bigger_numbers[left_num] = set()
            bigger_numbers[left_num].add(right_num)
            if right_num not in smaller_numbers:
                smaller_numbers[right_num] = set()
            smaller_numbers[right_num].add(left_num)

            if right_num not in bigger_numbers:
                bigger_numbers[right_num] = set()
            if left_num not in smaller_numbers:
                smaller_numbers[left_num] = set()

        updates = []
        for update_str in updates_str:
            updates.append([int(num) for num in update_str.split(",")])

        return updates, bigger_numbers, smaller_numbers


def part1(filename):

    updates, bigger_numbers, smaller_numbers = parse_input(filename)

    res = 0
    for update in updates:
        curr_left = set()
        curr_right = set(update)
        valid = True
        for el in update:
            curr_right.remove(el)
            if len(curr_left - smaller_numbers[el]) > 0:
                valid = False
                break
            if len(curr_right - bigger_numbers[el]) > 0:
                valid = False
                break
            curr_left.add(el)
        if valid:
            res += update[len(update) // 2]

    print("Part 1 result: ", res)


def part2(filename):

    updates, bigger_numbers, smaller_numbers = parse_input(filename)

    ## iterative way to find all bigger numbers
    # for num, bigger_nums in bigger_numbers.items():
    #     stack = list(bigger_nums)
    #     visited = set()
    #     while stack:
    #         curr = stack.pop()
    #         if curr in visited:
    #             continue
    #         visited.add(curr)
    #         bigger_numbers[num] = bigger_numbers[num] + bigger_numbers[curr]
    #         stack.extend(bigger_numbers[curr] - visited)

    ## recursive way to find all bigger numbers
    completed_bigger_numbers = set()

    def complete_all_bigger_nums(num):
        if num not in completed_bigger_numbers:
            bigger_numbers[num].union(
                *(complete_all_bigger_nums(n) for n in bigger_numbers[num])
            )
            completed_bigger_numbers.add(num)
        return bigger_numbers[num]

    res = 0
    for update in updates:
        curr_left = set()
        curr_right = set(update)
        valid = True
        for el in update:
            curr_right.remove(el)
            if len(curr_left - smaller_numbers[el]) > 0:
                valid = False
                break
            if len(curr_right - bigger_numbers[el]) > 0:
                valid = False
                break
            curr_left.add(el)
        if not valid:
            # custom sorting where bigger means numbers that appear in bigger_numbers[x]
            sorted_update = sorted(
                update,
                key=cmp_to_key(lambda x, y: (-1 if y in bigger_numbers[x] else 1)),
                reverse=True,
            )
            res += sorted_update[len(sorted_update) // 2]

    print("Part 2 result: ", res)


part1("test.txt")
part1("input.txt")
part2("test.txt")
part2("input.txt")

File: day5/day5.exs
defmodule Day5 do
  @moduledoc """
  Module for solving Day 5 of the challenge.
  """
  def parse_input(filename) do
    input = File.read!(filename)
    [rules, updates] = String.split(input, "\n\n")

    rules = String.split(rules, "\n")
    updates_str = String.split(updates, "\n")

    {bigger_numbers, smaller_numbers} = Enum.reduce(rules, {%{}, %{}}, fn rule, {bigger, smaller} ->
      [left_num, right_num] =
        rule
        |> String.split("|")
        |> Enum.map(&String.to_integer/1)

      bigger = Map.update(bigger, left_num, MapSet.new([right_num]), fn set ->
        MapSet.put(set, right_num)
      end)
      bigger = Map.put_new(bigger, right_num, MapSet.new())

      smaller = Map.update(smaller, right_num, MapSet.new([left_num]), fn set ->
        MapSet.put(set, left_num)
      end)
      smaller = Map.put_new(smaller, left_num, MapSet.new())

      {bigger, smaller}
    end)

    updates = Enum.map(updates_str, fn update -> 
      update
      |> String.split(",")
      |> Enum.map(&String.to_integer/1) 
    end)

    {updates, bigger_numbers, smaller_numbers}
  end

  defp check_validity(update, left, right, smaller_numbers, bigger_numbers) do
    Enum.reduce_while(update, {true, left, right}, fn el, {_, curr_left, curr_right} ->

        curr_right = MapSet.delete(curr_right, el)

        if MapSet.size(MapSet.difference(curr_left, smaller_numbers[el])) > 0 do
          {:halt, {false, curr_left, curr_right}}
        else if MapSet.size(MapSet.difference(curr_right, MapSet.new(bigger_numbers[el]))) > 0 do
          {:halt, {false, curr_left, curr_right}}
        else
          curr_left = MapSet.put(curr_left, el)
          {:cont, {true, curr_left, curr_right}}
        end
      end
        
    end) |> elem(0)
  end

  def part1(filename) do
    {updates, bigger_numbers, smaller_numbers} = parse_input(filename)

    res =
      Enum.reduce(updates, 0, fn update, res ->
        left = MapSet.new()
        right = MapSet.new(update)
        valid = check_validity(update, left, right, smaller_numbers, bigger_numbers)

        if valid do
          res + Enum.at(update, div(Enum.count(update), 2))
        else
          res
        end
      end)

    IO.puts("Result part1: #{res}")
  end

end

# tests
Day5.part1("test.txt")
# Day5.part2("test.txt")

# solve
Day5.part1("input.txt")
# Day5.part2("input.txt")
